pkg/operator/controller.go

```diff
@@ -1,481 +1,302 @@
-	controller.clients.Core, err = coreset.NewForConfig(controller.kubeconfig)
+func (r *TunedReconciler) defaultTunedRequests(nodeObj client.Object) []reconcile.Request {
+	var requests []reconcile.Request
+	defaultTuned, err := r.getDefaultTuned(context.TODO())
 	if err != nil {
-		return nil, err
+		return nil
 	}
+	requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+		Namespace: defaultTuned.Namespace,
+		Name:      defaultTuned.Name,
+	}})
+	// TODO - consider using syncprofile(node) here with the given node name
+	// this might be an abouse for a enqueue map function but might produce better performance
+	// the caveat is that this operation might not be in full sync with the reconciler.
+	return requests
+}
 
-	// DaemonSet
-	controller.clients.Apps, err = appsset.NewForConfig(controller.kubeconfig)
-	if err != nil {
-		return nil, err
+func (r *TunedReconciler) podLabelsToTuned(podObj client.Object) []reconcile.Request {
+	var requests []reconcile.Request
+	if r.state.enablePodEvents == false {
+		return nil
 	}
 
-	// ClusterOperator
-	controller.clients.ConfigClientSet, err = configclientset.NewForConfig(controller.kubeconfig)
+	defaultTuned, err := r.getDefaultTuned(context.TODO())
 	if err != nil {
-		return nil, err
+		return nil
 	}
-
-	return controller, nil
+	requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+		Namespace: defaultTuned.Namespace,
+		Name:      defaultTuned.Name,
+	}})
+	return requests
 }
 
-// eventProcessor is a long-running function that will continually
-// read and process a message on the workqueue.
-func (c *Controller) eventProcessor() {
-	for {
-		// Wait until there is a new item in the working queue
-		obj, shutdown := c.workqueue.Get()
-		if shutdown {
-			return
-		}
+// Reconcile reads that state of the cluster for a Tuned object and makes changes based on the state read
+// and what is in the Tuned.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *TunedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.Infof("Reconciling Tuned")
 
-		klog.V(2).Infof("got event from workqueue")
-		func() {
-			defer c.workqueue.Done(obj)
-			var workqueueKey wqKey
-			var ok bool
+	//TODO - consider using  custom req.NamespacedName for special events such as node deletion.
+	// see if this is not an abuse.
 
-			if workqueueKey, ok = obj.(wqKey); !ok {
-				c.workqueue.Forget(obj)
-				klog.Errorf("expected wqKey in workqueue but got %#v", obj)
-				return
-			}
+	// Fetch the Tuned instance
+	tunedInstance := &tunedv1.Tuned{}
+	err := r.Get(ctx, req.NamespacedName, tunedInstance)
+	if err != nil {
+		// for deletion events perform all the needed syncs as a regular event.
+		if !errors.IsNotFound(err) {
```

Comment:
if err equals to `IsNotFound` we should exit the reconcile loop early

Reply:
Since we are using a tree based structure for tuned -> profiles, when a tuned CR is deleted we need to go through the whole sync to update the relevant artifacts.  you will notice that the sync work is mostly done against the default tuned CR and not the reconciled CR itself (it is just the trigger) 
We are Listing all the tuned CRs to get the recommended profiles down the pipeline here .
I'm not entirely happy this behavior and open for suggestions.

Reply:
ok works for me